### PR TITLE
[WIP] make repr formatting narrower 

### LINKED ISSUE #857: Consider implementing Rich Repr Protocol for classes in hf_api
**Is your feature request related to a problem? Please describe.**

Currently, the `__repr__`'s for classes in the `hf_api` module are nice and compact but can require a lot of side-scrolling to view all of the information. For example, the `__repr__` for `ModelInfo` looks like:

 ```
ModelInfo: {
	modelId: flyswot/convnext-tiny-224_flyswot
	sha: c6d4b2138e10efeafef8f5305ce16270ca583618
	lastModified: 2022-04-05T16:08:35.000Z
	tags: ['pytorch', 'convnext', 'image-classification', 'dataset:image_folder', 'transformers', 'generated_from_trainer', 'model-index']
	pipeline_tag: image-classification
	siblings: [ModelFile(rfilename='.gitattributes'), ModelFile(rfilename='.gitignore'), ModelFile(rfilename='README.md'), ModelFile(rfilename='config.json'), ModelFile(rfilename='preprocessor_config.json'), ModelFile(rfilename='pytorch_model.bin'), ModelFile(rfilename='training_args.bin')]
	config: {'architectures': ['ConvNextForImageClassification'], 'model_type': 'convnext'}
	id: flyswot/convnext-tiny-224_flyswot
	author: flyswot
	private: False
	downloads: 999
	library_name: transformers
	model-index: [{'name': 'convnext-tiny-224_flyswot', 'results': [{'task': {'name': 'Image Classification', 'type': 'image-classification'}, 'dataset': {'name': 'image_folder', 'type': 'image_folder', 'args': 'default'}, 'metrics': [{'name': 'F1', 'type': 'f1', 'value': 0.9756290792360154}]}]}]
	cardData: {'tags': ['generated_from_trainer'], 'datasets': ['image_folder'], 'metrics': ['f1'], 'model-index': [{'name': 'convnext-tiny-224_flyswot', 'results': [{'task': {'name': 'Image Classification', 'type': 'image-classification'}, 'dataset': {'name': 'image_folder', 'type': 'image_folder', 'args': 'default'}, 'metrics': [{'name': 'F1', 'type': 'f1', 'value': 0.9756290792360154}]}]}]}
}
```

**Describe the solution you'd like**

[`rich`](https://github.com/Textualize/rich/) is a:

> Python library for rich text and beautiful formatting in the terminal.

One of the features `rich` offers its own `print`. An example of this 

Screenshot 2022-04-28 at 14 17 21

However: 
> Rich is able to syntax highlight any output, but the formatting is restricted to built-in containers, dataclasses, and other objects Rich knows about, such as objects generated by the [attrs](https://www.attrs.org/en/stable/) library. To add Rich formatting capabilities to custom objects, you can implement the rich repr protocol. -[source](https://rich.readthedocs.io/en/latest/pretty.html#rich-repr-protocol)

It might be nice to add this for the container classes under `hf_api`. This requires implementing `__rich_rep__`. 

As an example, I have created a simple version of this for `ModelInfo`:

```python
def __rich_repr__(self:ModelInfo):
    for key, val in self.__dict__.items():
        yield key, val
```

This results in the following when using `rich`'s version of `print`:

```python
ModelInfo(
    modelId='flyswot/convnext-tiny-224_flyswot',
    sha='c6d4b2138e10efeafef8f5305ce16270ca583618',
    lastModified='2022-04-05T16:08:35.000Z',
    tags=[
        'pytorch',
        'convnext',
        'image-classification',
        'dataset:image_folder',
        'transformers',
        'generated_from_trainer',
        'model-index'
    ],
    pipeline_tag='image-classification',
    siblings=[
        ModelFile(rfilename='.gitattributes'),
        ModelFile(rfilename='.gitignore'),
        ModelFile(rfilename='README.md'),
        ModelFile(rfilename='config.json'),
        ModelFile(rfilename='preprocessor_config.json'),
        ModelFile(rfilename='pytorch_model.bin'),
        ModelFile(rfilename='training_args.bin')
    ],
    config={'architectures': ['ConvNextForImageClassification'], 'model_type': 'convnext'},
    id='flyswot/convnext-tiny-224_flyswot',
    author='flyswot',
    private=False,
    downloads=999,
    library_name='transformers',
    model-index=[
        {
            'name': 'convnext-tiny-224_flyswot',
            'results': [
                {
                    'task': {
                        'name': 'Image Classification',
                        'type': 'image-classification'
                    },
                    'dataset': {
                        'name': 'image_folder',
                        'type': 'image_folder',
                        'args': 'default'
                    },
                    'metrics': [{'name': 'F1', 'type': 'f1', 'value': 0.9756290792360154}]
                }
            ]
        }
    ],
    cardData={
        'tags': ['generated_from_trainer'],
        'datasets': ['image_folder'],
        'metrics': ['f1'],
        'model-index': [
            {
                'name': 'convnext-tiny-224_flyswot',
                'results': [
                    {
                        'task': {
                            'name': 'Image Classification',
                            'type': 'image-classification'
                        },
                        'dataset': {
                            'name': 'image_folder',
                            'type': 'image_folder',
                            'args': 'default'
                        },
                        'metrics': [
                            {'name': 'F1', 'type': 'f1', 'value': 0.9756290792360154}
                        ]
                    }
                ]
            }
        ]
    }
)
```

It also comes with nice colour highlighting:

Screenshot 2022-04-28 at 14 23 01

There is scope for making this nicer, but even though this version is much longer, it does allow everything to be visible, even on narrow displays. 

One of the reasons I think this could be a candidate for inclusion is that:

> you can add __rich_repr__ methods to third-party libraries without including Rich as a dependency. If Rich is not installed, then nothing will break. Hopefully more third-party libraries will adopt Rich repr methods in the future.

This means that implementing a `__rich__` repr doesn't require this library or end-users, to have `rich` installed. If they like the features of `rich`, they get the benefits, but it's not forced on anyone. 

**Describe alternatives you've considered**

Not doing this! Although I find even the naive `__rich_repr__` quite nice, this is a matter of personal preference. There is also a potential scope creep with implementing `rich_repr__` across more and more code. However, I think the container classes in `hf_api` are particularly good candidates because it contains information you might want to inspect on a terminal/print out in a notebook etc. 

**Additional context**

If this is something that is deemed worth exploring more I would be happy to make an initial pull request to begin implementing this.

### PR DESCRIPTION
**Pull request by davanstrien**
## What this PR does?

This PR aims to make the `__repr__`s in `hf_api.py` easier to view on a narrow screen i.e. to favour a longer `__repr__` for a class over a short but very wide `__repr__`. This PR closes #857  

## Intended outcome 

As discussed in #857, the goal here is to have a `__repr__` that is easy to view on a narrow screen without horizontal scrolling. 

I have tried to make sure the approach taken doesn't do anything esoteric and doesn't add too much additional code to the `__repr__`. I have tried to keep the format of the `__repr__` as similar to what is there now, except for making it narrower. 

## Approach taken 

For now, I have just changed the `__repr__` for `ModelInfo` to get initial feedback. The `__repr__` now uses `pprint.pformat` to offload the formatting work for the attributes inside the class. At the moment, the `__repr__`s for all the classes in `hf_api.py` show all the attributes, but it would also be possible to hide some before passing to `pformat` if needed. I chose this approach since many people working with python will be at least vaguely familiar with pprint so hopefully, the code is quite easily understood. 

My suggestion, if this approach/format looks okay, would be to add this to all the classes in `hf_api`. There is also the option of tweaking the `width` parameter used by `pformat`. 

An example of the output given by the old and new `__repr__` are shown below. 

### old

```python
ModelInfo: {
	modelId: flyswot/convnext-tiny-224_flyswot
	sha: c6d4b2138e10efeafef8f5305ce16270ca583618
	lastModified: 2022-04-05T16:08:35.000Z
	tags: ['pytorch', 'convnext', 'image-classification', 'dataset:image_folder', 'transformers', 'generated_from_trainer', 'model-index']
	pipeline_tag: image-classification
	siblings: [RepoFile(rfilename='.gitattributes'), RepoFile(rfilename='.gitignore'), RepoFile(rfilename='README.md'), RepoFile(rfilename='config.json'), RepoFile(rfilename='preprocessor_config.json'), RepoFile(rfilename='pytorch_model.bin'), RepoFile(rfilename='training_args.bin')]
	private: False
	author: flyswot
	config: {'architectures': ['ConvNextForImageClassification'], 'model_type': 'convnext'a
	id: flyswot/convnext-tiny-224_flyswot
	downloads: 984
	library_name: transformers
	model-index: [{'name': 'convnext-tiny-224_flyswot', 'results': [{'task': {'name': 'Image Classification', 'type': 'image-classification'}, 'dataset': {'name': 'image_folder', 'type': 'image_folder', 'args': 'default'}, 'metrics': [{'name': 'F1', 'type': 'f1', 'value': 0.9756290792360154}]}]}]
	cardData: {'tags': ['generated_from_trainer'], 'datasets': ['image_folder'], 'metrics': ['f1'], 'model-index': [{'name': 'convnext-tiny-224_flyswot', 'results': [{'task': {'name': 'Image Classification', 'type': 'image-classification'}, 'dataset': {'name': 'image_folder', 'type': 'image_folder', 'args': 'default'}, 'metrics': [{'name': 'F1', 'type': 'f1', 'value': 0.9756290792360154}]}]}]}
	transformersInfo: {'auto_model': 'AutoModelForImageClassification', 'pipeline_tag': 'image-classification', 'processor': 'AutoFeatureExtractor'}
}
```
### new 

```python
ModelInfo: { 
        {'modelId': 'flyswot/convnext-tiny-224_flyswot',
         'sha': 'c6d4b2138e10efeafef8f5305ce16270ca583618',
         'lastModified': '2022-04-05T16:08:35.000Z',
         'tags': ['pytorch',
                  'convnext',
                  'image-classification',
                  'dataset:image_folder',
                  'transformers',
                  'generated_from_trainer',
                  'model-index'],
         'pipeline_tag': 'image-classification',
         'siblings': [RepoFile(rfilename='.gitattributes'),
                      RepoFile(rfilename='.gitignore'),
                      RepoFile(rfilename='README.md'),
                      RepoFile(rfilename='config.json'),
                      RepoFile(rfilename='preprocessor_config.json'),
                      RepoFile(rfilename='pytorch_model.bin'),
                      RepoFile(rfilename='training_args.bin')],
         'private': False,
         'author': 'flyswot',
         'config': {'architectures': ['ConvNextForImageClassification'],
                    'model_type': 'convnext'},
         'id': 'flyswot/convnext-tiny-224_flyswot',
         'downloads': 984,
         'library_name': 'transformers',
         'model-index': [{'name': 'convnext-tiny-224_flyswot',
                          'results': [{'task': {'name': 'Image Classification',
                                                'type': 'image-classification'},
                                       'dataset': {'name': 'image_folder',
                                                   'type': 'image_folder',
                                                   'args': 'default'},
                                       'metrics': [{'name': 'F1',
                                                    'type': 'f1',
                                                    'value': 0.9756290792360154}]}]}],
         'cardData': {'tags': ['generated_from_trainer'],
                      'datasets': ['image_folder'],
                      'metrics': ['f1'],
                      'model-index': [{'name': 'convnext-tiny-224_flyswot',
                                       'results': [{'task': {'name': 'Image Classification',
                                                             'type': 'image-classification'},
                                                    'dataset': {'name': 'image_folder',
                                                                'type': 'image_folder',
                                                                'args': 'default'},
                                                    'metrics': [{'name': 'F1',
                                                                 'type': 'f1',
                                                                 'value': 0.9756290792360154}]}]}]},
         'transformersInfo': {'auto_model': 'AutoModelForImageClassification',
                              'pipeline_tag': 'image-classification',
                              'processor': 'AutoFeatureExtractor'}}
	}
```

### CODE REVIEW

**Comment by davanstrien** [comment]
cc @nateraw @adrinjalali @LysandreJik

File: src/huggingface_hub/hf_api.py [hunk 1]
```diff
@@ -6,16 +6,18 @@
 # You may obtain a copy of the License at
 #
 #     http://www.apache.org/licenses/LICENSE-2.0
-#
+# §
```
Comment (julien-c) [inline src/huggingface_hub/hf_api.py:9]: extraneous change?

Comment (davanstrien) [reply src/huggingface_hub/hf_api.py]: Yes, typo 😬 fixed that now. Thanks for spotting

**Comment by Wauplin** [comment]
Hi @davanstrien thanks for your PR. I want to follow-up on it. Is it still a Draft/WIP work or it's ready to review/merge ? 
(cc @julien-c @LysandreJik as well)

**Comment by davanstrien** [comment]
> Hi @davanstrien thanks for your PR. I want to follow-up on it. Is it still a Draft/WIP work or it's ready to review/merge ? (cc @julien-c @LysandreJik as well)

If everyone is happy with the suggested approach then I would suggest adding this updated `_repr__` to all the classes in `hf_api.py`. Once that is done this PR should be ready for final review/to be merged.

**Comment by Wauplin** [comment]
I quickly looked at `hf_api.py` and as you suggested it makes sense to adapt the `__repr__`  for all of them (`RepoFile` `ModelInfo`, `DatasetInfo`, `SpaceInfo` and `MetricInfo`). I think it would also be good to group everything in a single abstract class (implementing only `__repr__`) and give it by inheritance instead of the current duplicated code.

@davanstrien Since you already started working on it, would you mind taking care of it ? Thanks in advance :)

**Comment by davanstrien** [comment]
> I quickly looked at `hf_api.py` and as you suggested it makes sense to adapt the `__repr__` for all of them (`RepoFile` `ModelInfo`, `DatasetInfo`, `SpaceInfo` and `MetricInfo`). I think it would also be good to group everything in a single abstract class (implementing only `__repr__`) and give it by inheritance instead of the current duplicated code.
> 
> @davanstrien Since you already started working on it, would you mind taking care of it ? Thanks in advance :)

Sure, I'm super busy at the moment so I likely won't be able to look at this until mid-September.

**Comment by Wauplin** [comment]
Ok thanks, that's fine 👍

**Comment by davanstrien** [comment]
@Wauplin I've used a Mixin to implement this new `__repr__` format. Let me know if you have something else in mind :)
I don't think the failing tests are because of these changes but let me know if I need to do anything to fix these.

**Review by Wauplin — APPROVED** [review]
Thanks for completing this @davanstrien ! Failing tests are indeed not related. I'm merging the PR :)

**Pull request merged**